// ionic4-auto-complete/esm2015/index.js
var AutoCompleteModule_1;
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { IonicModule } from '@ionic/angular';
import { AutoCompleteComponent } from './auto-complete/auto-complete.component';
import { BoldPrefix } from './bold-prefix.pipe';
export * from './auto-complete-options.model';
export * from './auto-complete/auto-complete.component';
export * from './bold-prefix.pipe';
let AutoCompleteModule = AutoCompleteModule_1 = class AutoCompleteModule {
    static forRoot() {
        return {
            ngModule: AutoCompleteModule_1,
            providers: []
        };
    }
};
AutoCompleteModule = AutoCompleteModule_1 = tslib_1.__decorate([
    NgModule({
        declarations: [
            AutoCompleteComponent,
            BoldPrefix
        ],
        exports: [
            AutoCompleteComponent,
            BoldPrefix
        ],
        imports: [
            CommonModule,
            FormsModule,
            IonicModule
        ]
    })
], AutoCompleteModule);
export { AutoCompleteModule };
